modules/jsonld/src/main/scala/ch/epfl/bluebrain/nexus/rdf/jsonld/JsonLd.scala

```diff
@@ -0,0 +1,147 @@
+      }
+    inner(List.empty, contextValue(json)).map(flattened => replaceContext(json, Json.obj("@context" -> flattened)))
+  }
+
+  /**
+    * @return a new Json with the values of all the ''@context'' keys
+    */
+  def contextValue(json: Json): Json =
+    (json.asObject, json.asArray) match {
+      case (Some(jObj), _) if jObj.nonEmpty =>
+        val context = jObj("@context").getOrElse(Json.obj())
+        jObj.remove("@context").values.foldLeft(context)((acc, c) => merge(acc, contextValue(c)))
+      case (_, Some(arr)) if arr.nonEmpty =>
+        arr.foldLeft(Json.obj())((acc, c) => merge(acc, contextValue(c)))
+      case _ =>
+        Json.obj()
+    }
+
+  /**
+    * Replaces the @context value from the provided json to the one in ''that'' json
+    *
+    * @param json the primary json
+    * @param that the json with a @context to override the @context in the provided ''json''
+    */
+  def replaceContext(json: Json, that: Json): Json =
+    removeNestedKeys(json, "@context") deepMerge Json.obj("@context" -> contextValue(that))
+
+  /**
+    * Removes the provided keys from everywhere on the json.
+    *
+    * @param json the json
+    * @param keys list of ''keys'' to be removed from the top level of the ''json''
+    * @return the original json without the provided ''keys''
+    */
+  def removeNestedKeys(json: Json, keys: String*): Json = {
+    def inner(obj: JsonObject): JsonObject =
+      JsonObject.fromIterable(
+        obj.filterKeys(!keys.contains(_)).toVector.map { case (k, v) => k -> removeNestedKeys(v, keys: _*) }
+      )
+    json.arrayOrObject[Json](
+      json,
+      arr => Json.fromValues(removeEmpty(arr.map(j => removeNestedKeys(j, keys: _*)))),
+      obj => inner(obj).asJson
+    )
+  }
+
+  private def removeEmpty(arr: Seq[Json]): Seq[Json] =
+    arr.filter(j => j != Json.obj() && j != Json.fromString("") && j != Json.arr())
+
+  private def merge(json: Json, that: Json): Json =
+    (json.asArray, that.asArray, json.asString, that.asString) match {
+      case (Some(arr), Some(thatArr), _, _) => Json.arr(removeEmpty(arr ++ thatArr): _*)
+      case (_, Some(thatArr), _, _)         => Json.arr(removeEmpty(json +: thatArr): _*)
+      case (Some(arr), _, _, _)             => Json.arr(removeEmpty(arr :+ that): _*)
+      // $COVERAGE-OFF$
+      case (_, _, Some(str), Some(thatStr)) => Json.arr(removeEmpty(Seq(str.asJson, thatStr.asJson)): _*)
+      case (_, _, Some(str), _)             => Json.arr(removeEmpty(Seq(str.asJson, that)): _*)
+      case (_, _, _, Some(thatStr))         => Json.arr(removeEmpty(Seq(json, thatStr.asJson)): _*)
+      // $COVERAGE-ON$
+      case _ => json deepMerge that
+    }
+
+  /**
+    * Exception signalling an error during context resolution.
+    * @param msg error message
+    */
+  sealed abstract class ContextResolutionException(val msg: String) extends Exception with Product with Serializable {
```

Comment:
We now have three types of "errors" in the RDF lib: DecodingFailure, DerivationError and now ContextResolutionException. We should stick with one of Failure, Error and Exception suffixes. I would vote with Error, being shorter. Thoughts?

Reply:
yes, it makes sense to make it consistent, I think Error makes sense, I'll update it